### PR TITLE
[pytree] support collections.defaultdict type for Python pytree

### PR DESCRIPTION
**Pull request by XuehaiPan**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #114858
* #113258
* #113257
* #113256
* __->__ #113255
* #112485

cc @zou3519

### CODE REVIEW

File: torch/utils/_pytree.py [hunk 1]
```diff
@@ -326,22 +328,59 @@ def _namedtuple_deserialize(dumpable_context: DumpableContext) -> Context:
...
+
+
+def _defaultdict_unflatten(
+    values: Iterable[Any], context: Context
+) -> DefaultDict[Any, Any]:
+    default_factory, keys = context
+    return defaultdict(default_factory, zip(keys, values))
+
+
+def _defaultdict_serialize(context: Context) -> DumpableContext:
```
Comment (zou3519) [inline torch/utils/_pytree.py:358]: Can you add a test for _defaultdict_serialize/_defaultdict_deserialize?

**Review by zou3519 — COMMENTED** [review]
Needs a test for serializing/deserializing defaultdict, otherwise, LGTM

File: torch/utils/_pytree.py (hunk 1 above)
Comment (XuehaiPan) [reply torch/utils/_pytree.py:359]: I add a serialization test case for `defaultdict`.

**Review by zou3519 — APPROVED** [review]

**Pull request closed without merging**